libraries/chain/kv_context_rocksdb.cpp

```diff
@@ -0,0 +1,315 @@
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         return kv_it_status();
+      }
+
+      kv_it_stat kv_it_prev(uint32_t* found_key_size, uint32_t* found_value_size) override {
+         EOS_ASSERT(!kv_it.is_erased(), kv_bad_iter, "Iterator to erased element");
+         try {
+            try {
+               --kv_it;
+               get_current_key_value_sizes(found_key_size, found_value_size);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         return kv_it_status();
+      }
+
+      kv_it_stat kv_it_lower_bound(const char* key, uint32_t size, uint32_t* found_key_size, uint32_t* found_value_size) override {
+         try {
+            try {
+               kv_it.lower_bound(key, size);
+               get_current_key_value_sizes(found_key_size, found_value_size);
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         return kv_it_status();
+      }
+
+      kv_it_stat kv_it_key(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) override {
+         EOS_ASSERT(!kv_it.is_erased(), kv_bad_iter, "Iterator to erased element");
+
+         std::optional<b1::chain_kv::key_value> kv;
+         try {
+            try {
+               kv = kv_it.get_kv();
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         if (kv) {
+            if (offset < kv->key.size())
+               memcpy(dest, kv->key.data() + offset, std::min((size_t)size, kv->key.size() - offset));
+            actual_size = kv->key.size();
+            return kv_it_stat::iterator_ok;
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+      kv_it_stat kv_it_value(uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size) override {
+         EOS_ASSERT(!kv_it.is_erased(), kv_bad_iter, "Iterator to erased element");
+
+         std::optional<b1::chain_kv::key_value> kv;
+         try {
+            try {
+               kv = kv_it.get_kv();
+            }
+            FC_LOG_AND_RETHROW()
+         }
+         CATCH_AND_EXIT_DB_FAILURE()
+
+         if (kv) {
+            if (offset < kv->value.size())
+               memcpy(dest, kv->value.data() + offset, std::min((size_t)size, kv->value.size() - offset));
+            actual_size = kv->value.size();
+            return kv_it_stat::iterator_ok;
+         } else {
+            actual_size = 0;
+            return kv_it_stat::iterator_end;
+         }
+      }
+
+    private:
+      void get_current_key_value_sizes(uint32_t* found_key_size, uint32_t* found_value_size) {
+         if (kv_it.is_valid()) {
+            // Not end or at an erased kv pair
+            auto kv = kv_it.get_kv();
+
+            // kv is always non-null due to the check of is_valid()
+            *found_key_size = kv->key.size();
```

Comment:
Who is responsible for the null check on `found_key_size` and `found_value_size`?